### ISSUE TITLE
Receiver c534 not handled correctly (with some proposed fixes)

### ISSUE THREAD

**Issue by pfps** [pwr-Solaar/Solaar#598]
**Information**

- Solaar version: 1.0.1
- Distribution: Fedora
- Kernel version (ex. `uname -srmo`): Linux 5.3.14-300.fc31.x86_64 x86_64 GNU/Linux
- Output of `solaar show`:
```
Unifying Receiver
  Device path  : /dev/hidraw6
  USB id       : 046d:c534
  Serial       : 0
    Firmware   : 29.01.B0016
  Has 2 paired device(s) out of a maximum of 6.
  Notifications: wireless, software present (0x000900)

  1: Wireless Keyboard MK270
     Codename     : MK270
     Kind         : keyboard
     Wireless PID : 4023
     Protocol     : HID++ 2.0
     Polling rate : 20 ms (50Hz)
     Serial number: 90A794A3
     Battery: unknown (device is offline).

  2: Wireless Mouse M185 new
     Codename     : M185n
     Kind         : mouse
     Wireless PID : 4054
     Protocol     : HID++ 4.5
     Polling rate : 8 ms (125Hz)
     Serial number: 6FAD17D5
     Battery: unknown (device is offline).
```

**Describe the bug**
Receiver c534 is a low-end receiver.  It can only pair one keyboard and one mouse, and only low-end keyboards (e.g., K270) and mice (e.g., M185).   It is incorrectly recognized as a unifying receiver that can pair six devices.

**To Reproduce**
1. solaar -D <c534 device> pair
2. Turn on a K520 keyboard.  No pairing happens.

1. solaar -D <c534 device> unpair 1
solaar: error: failed to unpair device Wireless Keyboard MK270: 1

1. solaar -D <c534 device> pair
2. Turn on a K270 keyboard or M185 mouse.
3. Keyboard replaces previous paired keyboard
4. No pairing reported
Pairing: turn your new device on (timing out in 20 seconds).
solaar: error: pairing failed: no device detected?

**Additional context**
There is special code for the device in receiver.py.  

Changing line 347 to
			self.max_devices = 2
provides the correct number of pairable devices and lists it as a Nano Receiver in the GUI.  Unfortunately this prevents repairing in the GUI.  I'll try to figure out a fix for this

Changing line 360 to
		self.may_unpair = self.product_id != 'c534' and self.write_register(_R.receiver_pairing) is None
provides the correct value for unpairing.

Adding lines 89 and 90 to lib/solaar/cli/pair.py as 
	elif receiver.product_id == 'c534':
		print ('This receiver does not report pairing results')
gives a better experience for CLI pairing.

PS: I don't like having c534 sprinkled in the code.  Can this be made a symbolic constant?

**Comment by FFY00** [comment]
> I don't like having c534 sprinkled in the code. Can this be made a symbolic constant?

Me neither but I'll accept it if it's too much trouble to implement it other way. Yes, send a patch and I'll review.

**Comment by pfps** [comment]
I'm looking at fixing up the GUI interface.  I'll submit a patch (or pull request) soon.

**Comment by pfps** [comment]
I have a few of these devices so I was trying to find out just how well they worked or didn't.
The CLI appears to work well, and the very minor changes above provide better information about the capabilities of the receiver and what is happening during pairing.  (Better messages would be better, but this was easy).
The GUI does not work well, as far as I can tell.  It can do one successful pair (actually an unpair and pair) but then no more pairing can be done on that receiver.  I can't do too much testing of this as I can't get the receiver back to its initial state.  I don't have current access to a Windows machine to see whether the Logitech software can fix the receiver. (I also can't confirm that the GUI pairing does this to a receiver because I have a limited number of these receivers.)
I'm tempted to prohibit pairing from the GUI, but there should be some notification in the GUI to use the CLI instead if this is done.  Better would be to figure out why the GUI is messing up the receiver.  When I get access to a Windows machine I'll try to reset a receiver.  If I can do that then I can do more testing.

**Comment by pfps** [comment]
Here are logs of solaar -dd pair on a receiver that has been bricked (bad) and one that hasn't (good).  There are some differences between responses from the two early on, but they are in fields that HID1.0 lists as reserved.  Is there a document on the newer protocols?

**Comment by FFY00** [comment]
My best guess is that the `0xc534` is one of the receivers that has a limited number of pairings (hardware limitation) and you reached the number of max pairings. Does the receiver have a unifying logo? Were you able to make a successful pair since?

**Comment by pfps** [comment]
This receiver does only have two pairings.  It can only pair a few devices.  It can't unpair (at least not in the usual way).  
On a working receiver, pairing can be opened.  A new mouse replaces the current one.  A new keyboard replaces the current one.  On the semi-broken receivers, pairing can't be opened - the response is a quick closure with an error message that is not in the HID++ 1.0 protocol.
I expect that there is something in the new HID++ protocols that controls pairing but I don't know where to look for documentation on this.
I'm very tempted to just turn off pairing in the GUI for this device for now.

**Comment by FFY00** [comment]
What I meant is that you can only pair a limited number of times. 

> On the semi-broken receivers, pairing can't be opened - the response is a quick closure with an error message that is not in the HID++ 1.0 protocol

That would indicate the receiver reached the maximum number of pairings, if my guess is correct.

**Comment by pfps** [comment]
Ooh - nasty!  Any documentation on this?

**Comment by FFY00** [comment]
Not yet, working on it.

**Comment by pfps** [comment]
I found what looks to be a connection count for c534 receivers.  It is the last byte of register 02 (which looks like a reasonable place).  The byte has the remaining number of connections plus 5.  (Why 5?  Who knows.)

Here is what is shown for two of my c534 receivers from hidconsole for the register.

	10 FF 8102 00020F	
	10 FF 8102 000205	

The first could be opened for pairing but the second could not.

A new connection for the first receiver changed the 0F to 0E.

But still   grrr, Grrr, GRRR!

**Comment by pfps** [comment]
PS:  Who was contributing new firmware for the Unifying receivers for fwupdmgr?  I was hoping that the c534 receivers would have an update.

**Comment by pfps** [comment]
I have some fixes for this receiver.  I'm hoping that my documentation changes can be pulled so that I can push these fixes - there are documentation changes for Nano receivers involved.

**Comment by pfps** [comment]
It occurs to me that this receiver may have the same hardware and software as a Unifying receiver with the only difference being the field in register 02 (and probably some other field giving the maximum number of paired devices - 0xB510 has a field that looks like it has the right value).  I wonder if there is there is a way to write these fields.

**Comment by pfps** [comment]
Closed by PR #674
